### PR TITLE
fix: bundle dynamic API routes correctly with split-api-routes

### PR DESCRIPTION
**Pull request by Skn0tt**
## Description

We discovered that dynamic API routes, e.g. `/api/shows/[id].js`, aren't bundled correctly. This is likely due to the square brackets being interpreted as special glob patterns, resulting in the underlying files not being included correctly.

This PR adds an e2e test that demonstrates the bug, and then fixes it by processing the glob patterns before passing them to the build process.

## Tests

You can test this change yourself like so:

1. TODO

## Relevant links (GitHub issues, etc.) or a picture of cute animal

https://github.com/netlify/pod-dev-foundations/issues/516

### CODE REVIEW

**Comment by Skn0tt** [comment]
Here's the failing test: https://github.com/netlify/next-runtime/actions/runs/5176735315/jobs/9325929677#step:10:236

**Comment by Skn0tt** [comment]
https://github.com/netlify/next-runtime/pull/2154/commits/d19f422e1f5f239c502c2e07353cb3c8f08a577b should make it green 🟢

**Review by nickytonline — COMMENTED** [review]
The fix looks good, but I'm trying to understand how our existing tests didn't fail as we have other dynamic routes. Is the API route splitting only enabled on the default demo site?

Also, it'd be good to test with and without the split API routes feature flag as it's not rolled out to everyone yet.

**Review by nickytonline — COMMENTED** [review]
The fix looks good, but I'm trying to understand how our existing tests didn't fail as we have other dynamic routes. Is the API route splitting only enabled on the default demo site?

Also, it'd be good to test with and without the split API routes feature flag as it's not rolled out to everyone yet.

**Comment by Skn0tt** [comment]
> The fix looks good, but I'm trying to understand how our existing tests didn't fail as we have other dynamic routes.

The bug only exists for API Routes, not for SSR routes, since we're currently only using the new bundling mechanism for API Routes. I don't think we tested any other dynamic API Routes before, or else we'd have caught this bug earlier.

> Is the API route splitting only enabled on the default demo site?

It's enabled in 4 demo sites, which I think are the sites that we're running e2e tests against: https://github.com/search?q=repo:netlify/next-runtime+NEXT_SPLIT_API_ROUTES&type=code

> Also, it'd be good to test with and without the split API routes feature flag as it's not rolled out to everyone yet.

With the way our e2e tests work, I don't see how we can do that. I think that's out-of-scope for this PR as well. I'd prefer that we get the split-api-routes work done quickly, and remove the featureflag before it becomes long-lived enough to warrant additional test cases.

File: packages/runtime/src/helpers/config.ts [hunk 1]
```diff
@@ -96,6 +96,12 @@ export const hasManuallyAddedModule = ({
   )
 /* eslint-enable camelcase */
 
+/**
+ * Transforms `/api/shows/[id].js` into `/api/shows/*id*.js`,
+ * so that the file `[id].js` is matched correctly.
+ */
+const escapeGlob = (path: string) => path.replace(/\[/g, '*').replace(/]/g, '*')
```
Comment (orinokai) [inline packages/runtime/src/helpers/config.ts:103]: i think this is going to match unwanted files, e.g. `grid.js`
is it possible to escape the path, e.g. `\[id\].js` ?

Comment (Skn0tt) [reply packages/runtime/src/helpers/config.ts:103]: I tried escaping the path, and glob patterns do allow escaping the brackets - but I couldn't get that to work properly with how ZISI interprets the globs. Maybe we could open a follow-up issue to investigate this.

Yes, this matches too many files - but I don't think that it'll be a big problem in practice.

Comment (Skn0tt) [reply packages/runtime/src/helpers/config.ts:103]: Good to know it's fixed in v9! I'd prefer the follow-up. Let's open an issue for this.

**Review by LekoArts — APPROVED** [review]

File: packages/runtime/src/helpers/config.ts (hunk 1 above)
Comment (LekoArts) [reply packages/runtime/src/helpers/config.ts:103]: Opened https://github.com/netlify/zip-it-and-ship-it/issues/1459

**Pull request merged**